### PR TITLE
kv graveler migration

### LINKED ISSUE #3573: Graveler Migration
**DoD**
Graveler migration in place. Esti tests to cover the migration (should be reenabled by the end of this test)

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #3573

### CODE REVIEW

File: esti/migrate_test.go [hunk 1]
```diff
@@ -542,3 +554,147 @@ func verifyUserPermissions(t *testing.T, ctx context.Context, repo, userType str
...
+func uploadFiles(t *testing.T, ctx context.Context, repo string, branch string, prefix string) {
+	for i := 0; i < migrateFileSize; i++ {
+		_, _, err := uploadFileRandomDataAndReport(ctx, repo, branch, prefix+strconv.Itoa(i), true)
+		if err != nil {
+			t.Error(err)
+		}
+	}
+}
+
+func verifyGravelerEntities(t *testing.T, ctx context.Context) {
```
Comment (itaiad200) [inline esti/migrate_test.go:653]: Please also validate the data of these fields and not just existence.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Cool 🆒 
Dropped some comments, also tried to guess why the test fails 🔎

File: pkg/graveler/ref/db_manager.go [hunk 2]
```diff
@@ -31,7 +31,7 @@ const (
 
 //nolint:gochecknoinits
 func init() {
-	kvpg.RegisterMigrate(packageName, Migrate, []string{"graveler_repositories"})
+	kvpg.RegisterMigrate(packageName, Migrate, []string{"graveler_repositories", "graveler_commits", "graveler_tags"})
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/db_manager.go]: done. thanks

File: pkg/graveler/ref/kv_migrate_test.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+		for b := 0; b < numBranches; b++ {
+			if err := mgr.CreateBranch(ctx, repoId, graveler.BranchID("branch_"+strconv.Itoa(b)),
+				graveler.Branch{StagingToken: graveler.StagingToken("test_token_" + strconv.Itoa(b))}); err != nil {
+				t.Fatalf("Create Branch: %s", err)
+			}
+		}
+		for c := 0; c < numCommits; c++ {
+			commitID, err := mgr.AddCommit(ctx, repoId,
+				graveler.Commit{Version: graveler.CurrentCommitVersion, Message: strconv.Itoa(c), Parents: graveler.CommitParents{"parent"}})
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/kv_migrate_test.go]: done

File: pkg/graveler/ref/kv_migrate_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	require.NoError(t, err)
+	defer repos.Close()
+	repoNum := 0
+	for repos.Next() {
+		repo := repos.Value()
+		kvr, err := kvMgr.GetRepository(ctx, repo.RepositoryID)
+		require.NoError(t, err)
+		dbr, err := dbMgr.GetRepository(ctx, repo.RepositoryID)
+		require.NoError(t, err)
+		require.Equal(t, kvr.StorageNamespace, dbr.StorageNamespace)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/kv_migrate_test.go]: done

File: pkg/graveler/ref/kv_migrate_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	defer branches.Close()
+	branchesNum := 0
+	for branches.Next() {
+		branch := branches.Value()
+		bkv, err := kvMgr.GetBranch(ctx, repoID, branch.BranchID)
+		require.NoError(t, err)
+
+		bdb, err := dbMgr.GetBranch(ctx, repoID, branch.BranchID)
+		require.NoError(t, err)
+		require.Equal(t, bkv.StagingToken, bdb.StagingToken)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/kv_migrate_test.go]: done

File: pkg/graveler/ref/kv_migrate_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	require.NoError(t, err)
+	defer commits.Close()
+	commitNum := 0
+	for commits.Next() {
+		commit := commits.Value()
+		kvc, err := kvMgr.GetCommit(ctx, repoID, commit.CommitID)
+		require.NoError(t, err)
+		dbm, err := dbMgr.GetCommit(ctx, repoID, commit.CommitID)
+		require.NoError(t, err)
+		require.Equal(t, kvc.Message, dbm.Message)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/kv_migrate_test.go]: done

File: pkg/graveler/ref/kv_migrate_test.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+		commitNum += 1
+
+		// verify tag
+		tag, err := kvMgr.GetTag(ctx, repoID, graveler.TagID("tag_"+commit.CommitID))
+		require.NoError(t, err)
+		require.Equal(t, *tag, commit.CommitID)
+	}
+	require.Equal(t, commitNum, numCommits+1)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/ref/kv_migrate_test.go:129]: done

File: esti/migrate_test.go [hunk 8]
```diff
@@ -542,3 +554,147 @@ func verifyUserPermissions(t *testing.T, ctx context.Context, repo, userType str
 		require.Equal(t, http.StatusUnauthorized, respDeleteBranch.StatusCode(), "unexpected success for %s - DeleteBranch", userType)
 	}
 }
+
+func testPreMigrateGraveler(t *testing.T) {
+	//create repository
+	ctx, _, repo := setupTest(t)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/migrate_test.go:557]: Every test has a unique repository.

File: esti/migrate_test.go [hunk 9]
```diff
@@ -542,3 +554,147 @@ func verifyUserPermissions(t *testing.T, ctx context.Context, repo, userType str
...
+
+	// upload files to main branch
+	uploadFiles(t, ctx, repo, mainBranch, "a/foo/")
+
+	// commit
+	commitResp, err := client.CommitWithResponse(ctx, repo, mainBranch, &api.CommitParams{}, api.CommitJSONRequestBody{
+		Message: "pre_commit",
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/migrate_test.go]: done

File: esti/migrate_test.go [hunk 10]
```diff
@@ -76,6 +78,14 @@ type userPermissions struct {
 	canDeleteBranch bool
 }
 
+type gravelerState struct {
+	Repo               string `json:"repo"`
+	CommitID           string `json:"commit_id"`
+	TagID              string `json:"tag_id"`
+	committedObjPath   string `json:"committed_obj_path"`
+	uncommittedObjPath string `json:"uncommitted_obj_path"`
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/migrate_test.go]: done, except the repo id that is used many times.

File: esti/migrate_test.go [hunk 11]
```diff
@@ -542,3 +554,147 @@ func verifyUserPermissions(t *testing.T, ctx context.Context, repo, userType str
...
+	require.Equal(t, http.StatusOK, unobjResp.StatusCode())
+}
+
+func verifyGravelerList(t *testing.T, ctx context.Context) {
+	// list pre- and post-entities
+	repoResp, err := client.ListRepositoriesWithResponse(ctx, &api.ListRepositoriesParams{})
+	require.NoError(t, err, "list repositories")
+	require.Equal(t, http.StatusOK, repoResp.StatusCode())
+	repoPayload := repoResp.JSON200
+	require.Equal(t, len(repoPayload.Results), 2)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply esti/migrate_test.go]: You're right. thanks

File: esti/migrate_test.go (hunk 1 above)
Comment (eden-ohana) [reply esti/migrate_test.go:647]: I'm validating it on the graveler migrate unit test

File: pkg/graveler/graveler.go [hunk 12]
```diff
@@ -305,7 +305,7 @@ type Branch struct {
 
 // BranchRecord holds BranchID with the associated Branch data
 type BranchRecord struct {
-	BranchID BranchID
+	BranchID BranchID `db:"id"`
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/graveler.go:308]: Because the get branch returns a branch, not a branch record (that includes a branch id)

File: pkg/graveler/graveler.go [hunk 13]
```diff
@@ -289,7 +289,7 @@ func (c Commit) Identity() []byte {
 
 // CommitRecord holds CommitID with the associated Commit data
 type CommitRecord struct {
-	CommitID CommitID `db:"id"`
+	CommitID CommitID
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:292]: Something went wrong here with the linting/comments

File: pkg/graveler/graveler.go [hunk 14]
```diff
@@ -1335,14 +1335,17 @@ func (g *KVGraveler) Get(ctx context.Context, repositoryID RepositoryID, ref Ref
...
+		if err != nil {
+			if !errors.Is(err, ErrNotFound) {
+				return nil, err
+			} // proceed to committed
+		} else { // In staging
+			if value == nil {
+				// tombstone - the entry was deleted on the branch => doesn't exist
+				return nil, ErrNotFound
+			}
+			return value, nil
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1347]: Did we have a bug here too? Let's open an issue, fix and add many tests to this very critical flow. The fact that it was discovered in a system test means we are lacking coverage.

File: pkg/graveler/ref/commit_record.go [hunk 15]
```diff
@@ -3,6 +3,8 @@ package ref
 import (
 	"time"
 
+	"google.golang.org/protobuf/types/known/timestamppb"
+
```
Comment (itaiad200) [inline pkg/graveler/ref/commit_record.go:7]: lint

File: pkg/graveler/ref/manager.go [hunk 16]
```diff
@@ -37,26 +37,28 @@ type CommitNode struct {
...
+	}
 	branch := &graveler.BranchData{
 		Id:           branchID.String(),
 		CommitId:     b.CommitID.String(),
 		StagingToken: b.StagingToken.String(),
-		SealedTokens: make([]string, 0),
-	}
-	for _, st := range b.SealedTokens {
-		branch.SealedTokens = append(branch.SealedTokens, st.String())
+		SealedTokens: sealedTokens,
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:61]: Notice that you changed the behaviour here from empty list to nil (since sealed tokens should be empty for old branches), is this expected?

File: pkg/graveler/graveler.go (hunk 13 above)
Comment (eden-ohana) [reply pkg/graveler/graveler.go:292]: done

File: pkg/graveler/ref/commit_record.go (hunk 15 above)
Comment (eden-ohana) [reply pkg/graveler/ref/commit_record.go:7]: done

File: pkg/graveler/graveler.go (hunk 14 above)
Comment (eden-ohana) [reply pkg/graveler/graveler.go:1347]: yes, @N-o-Z is handling it on a separate PR

File: pkg/graveler/ref/manager.go (hunk 16 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:61]: yes, it's because of the migration, it is saved as SealedTokens{nil} instead of SealedTokens{}, same for commit parents

File: pkg/graveler/graveler.go (hunk 14 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:1347]: We're not lacking in coverage - the unit tests are just old, there's a scenario that tests it, the mocks were just not configured correctly

File: clients/spark/build.sbt [hunk 17]
```diff
@@ -150,7 +150,9 @@ lazy val assemblySettings = Seq(
       .inProject,
     ShadeRule.rename("scala.collection.compat.**" -> "shadecompat.@1").inAll,
     ShadeRule.rename("okio.**" -> "okio.shaded.@0").inAll,
-    ShadeRule.rename("okhttp3.**" -> "okhttp3.shaded.@0").inAll
+    ShadeRule.rename("okhttp3.**" -> "okhttp3.shaded.@0").inAll,
+    ShadeRule.rename("reactor.netty.**" -> "shadereactor.netty.@1").inAll,
+    ShadeRule.rename("reactor.util.**" -> "shadereactor.util.@1").inAll
```
Comment (itaiad200) [inline clients/spark/build.sbt:155]: Unrelated changes from a rebase perhaps?

**Review by itaiad200 — APPROVED** [review]
🥳 🥳 🥳  LGTM 🥳 🥳 🥳 
Please make the tests pass with the latest version
